src/transformers/models/ict/modeling_ict.py

```diff
@@ -0,0 +1,869 @@
+        self.num_attention_heads = self.num_attention_heads - len(heads)
+        self.all_head_size = self.attention_head_size * self.num_attention_heads
+        self.pruned_heads = self.pruned_heads.union(heads)
+
+    def forward(
+        self, hidden_states, output_attentions: bool = False
+    ) -> Union[Tuple[torch.Tensor, torch.Tensor], Tuple[torch.Tensor]]:
+        mixed_query_layer = self.query(hidden_states)
+
+        key_layer = self.transpose_for_scores(self.key(hidden_states))
+        value_layer = self.transpose_for_scores(self.value(hidden_states))
+        query_layer = self.transpose_for_scores(mixed_query_layer)
+
+        # Take the dot product between "query" and "key" to get the raw attention scores.
+        attention_scores = torch.matmul(query_layer, key_layer.transpose(-1, -2))
+
+        attention_scores = attention_scores / math.sqrt(self.attention_head_size)
+
+        # Normalize the attention scores to probabilities.
+        attention_probs = nn.functional.softmax(attention_scores, dim=-1)
+
+        # This is actually dropping out entire tokens to attend to, which might
+        # seem a bit unusual, but is taken from the original Transformer paper.
+        attention_probs = self.attention_dropout(attention_probs)
+
+        context_layer = torch.matmul(attention_probs, value_layer)
+
+        context_layer = context_layer.permute(0, 2, 1, 3).contiguous()
+        new_context_layer_shape = context_layer.size()[:-2] + (self.all_head_size,)
+        context_layer = context_layer.view(new_context_layer_shape)
+
+        outputs = self.output(context_layer)
+        outputs = self.residual_dropout(outputs)
+
+        return (outputs, attention_probs) if output_attentions else (outputs,)
+
+
+class ICTLayer(nn.Module):
+    def __init__(self, config):
+        super().__init__()
+        num_embed = config.hidden_size
+        intermediate_size = config.intermediate_size
+        self.intermediate_act_fn = ACT2FN[config.activation_function]
+
+        self.ln_1 = nn.LayerNorm(num_embed)
+        self.attention = ICTSelfAttention(config)
+        self.ln_2 = nn.LayerNorm(num_embed)
+        self.mlp = nn.Sequential(
+            nn.Linear(num_embed, intermediate_size),
+            self.intermediate_act_fn,
+            nn.Linear(intermediate_size, num_embed),
+            nn.Dropout(config.resid_pdrop),
+        )
+
+    def forward(self, hidden_states, output_attentions: bool = False):
+        self_attention_outputs = self.attention(self.ln_1(hidden_states, output_attentions=output_attentions))
+        attention_output = self_attention_outputs[0]
+        outputs = self_attention_outputs[1:]  # add self attentions if we output attention weights
+
+        hidden_states = hidden_states + attention_output
+        hidden_states = hidden_states + self.mlp(self.ln_2(hidden_states))
+
+        outputs = (hidden_states,) + outputs
+
+        return outputs
+
+
+class ICTEncoder(nn.Module):
```

Comment:
This was previously in `ICTTransformerModel` but broke it out here. It reads so much easier — should've done a while ago 🤦‍♂️